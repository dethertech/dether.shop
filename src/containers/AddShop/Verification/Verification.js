import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import styled from 'styled-components';

import { ShopRecap, Button, TransactionFlow } from '../../../components';
import tr from '../../../translate';

import {
  addShop as addShopAction,
  resetTransaction as resetTransactionAction,
  fetchAll as fetchAllAction,
  openNotificationModal as openNotificationModalAction,
  displayShopWillAppearNotice as displayShopWillAppearNoticeAction,
} from '../../../actions';
import {
  addShop as addShopHelper,
  getLicenceShop,
  getShop,
} from '../../../helpers/ethereum';

import { notificationsTypes } from '../../../constants';

const ButtonsWrapper = styled.div`
  max-width: 48rem;
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-around;
  margin-top: 40px;
`;

export class Verification extends PureComponent {
  static propTypes = {
    pendingShop: PropTypes.shape({
      name: PropTypes.string.isRequired,
      cat: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      opening: PropTypes.string.isRequired,
      lat: PropTypes.string.isRequired,
      lng: PropTypes.string.isRequired,
      countryId: PropTypes.string.isRequired,
    }).isRequired,
    addShopToStore: PropTypes.func.isRequired,
    addShopToContract: PropTypes.func.isRequired,
    isTransactionPending: PropTypes.bool.isRequired,
    goBack: PropTypes.func.isRequired,
    fetchAll: PropTypes.func.isRequired,
    centerPosition: PropTypes.shape({}).isRequired,
    openNotificationModal: PropTypes.func.isRequired,
    dispatchDisplayShopWillAppearNoticeAction: PropTypes.func.isRequired,
  };

  state = {
    licencePrice: null,
    transactionSubmitted: false,
  };

  async componentWillMount() {
    const { pendingShop } = this.props;
    const licencePrice = await getLicenceShop(pendingShop.countryId);
    this.setState({ licencePrice });
  }

  checkTransaction = async () => {
    const {
      addShopToStore,
      resetTransaction,
      fetchAll,
      centerPosition,
      openNotificationModal,
      dispatchDisplayShopWillAppearNoticeAction,
    } = this.props;

    const shop = await getShop();
    if (shop) {
      this.setState({ transactionSubmitted: false });
      fetchAll(centerPosition);
      resetTransaction();
      addShopToStore(shop);
      openNotificationModal({
        type: notificationsTypes.SUCCESS,
        message: tr('notifications.shop_added'),
      });
      dispatchDisplayShopWillAppearNoticeAction();
      return true;
    }
  };

  handleError = () => {
    this.setState({ transactionSubmitted: false });
  };

  addShop = async () => {
    const { pendingShop, addShopToContract } = this.props;

    const transaction = await addShopToContract(pendingShop);
    return transaction;
  };

  submitTransaction = () => this.setState({ transactionSubmitted: true });

  render() {
    const { pendingShop, isTransactionPending, goBack } = this.props;
    const { licencePrice, transactionSubmitted } = this.state;
    if (isTransactionPending || transactionSubmitted) {
      return (
        <TransactionFlow
          isTransaction
          loader={{
            title: tr('add_form_verification.loader_title'),
            message: tr('add_form_verification.loader_add_message'),
          }}
          sendTransaction={this.addShop}
          onError={this.handleError}
          shop={pendingShop}
          checkTransaction={this.checkTransaction}
        />
      );
    }
    return (
      <Fragment>
        <ShopRecap
          licencePrice={licencePrice}
          {...pendingShop}
          address={pendingShop.addressString}
        />
        <ButtonsWrapper>
          <Button width="45%" theme="primary" onClick={this.submitTransaction}>
            {tr('add_form_verification.submit_button')}
          </Button>
          <Button width="45%" onClick={goBack}>
            {tr('add_form_verification.edit_button')}
          </Button>
        </ButtonsWrapper>
      </Fragment>
    );
  }
}

const mapStateToProps = ({ shop, map, transaction }) => ({
  pendingShop: shop.pendingShop,
  isTransactionPending: transaction.pending,
  centerPosition: map.centerPosition,
});

const mapDispatchToProps = dispatch => ({
  addShopToContract: addShopHelper,
  addShopToStore: bindActionCreators(addShopAction, dispatch),
  fetchAll: bindActionCreators(fetchAllAction, dispatch),
  resetTransaction: bindActionCreators(resetTransactionAction, dispatch),
  openNotificationModal: bindActionCreators(
    openNotificationModalAction,
    dispatch,
  ),
  dispatchDisplayShopWillAppearNoticeAction: bindActionCreators(
    displayShopWillAppearNoticeAction,
    dispatch,
  ),
});

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Verification),
);
