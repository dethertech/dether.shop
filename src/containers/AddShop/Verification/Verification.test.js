import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Verification } from './Verification';

jest.mock('axios', () => ({
  CancelToken: {
    source: () => ({
      token: '',
    }),
  },
}));
describe('Containers::AddShop::Verification', () => {

  const pendingShop = {
    lat: '30',
    lng: '3',
    addressString: '40 rue Citadelle Paris',
    countryId: 'fr',
    postalCode: '75010',
    name: 'Magasin',
    description: 'Super Magasin',
    cat: 'cool',
    opening: '0000000'
  };

  const props = {
    pendingShop,
    addShopToStore: () => null,
    addShopToContract: () => null,
    openNotificationModal: () => null,
    addAddShopTransaction: () => null,
    isTransactionPending: false,
    transactionHash: '0x23874grfb2yb38f4h23784',
    endTransaction: () => null,
    goBack: () => null,
    fetchAll: () => null,
    centerPosition: { lat: 30, lng: 3 },
    dispatchDisplayShopWillAppearNoticeAction: () => null,
  }

  const component = props => <Verification {...props} />

  it('should match shallow snapshop', () => {
    expect(shallow(component(props))).toMatchSnapshot('Verification::shallow');
  });
  it('should match mounted snapshop', () => {
    expect(toJson(component(props))).toMatchSnapshot('Verification::mount');
  });
});
