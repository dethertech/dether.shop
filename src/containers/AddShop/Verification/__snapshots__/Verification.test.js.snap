// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verification::mount 1`] = `undefined`;

exports[`Verification::shallow 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Verification
    addAddShopTransaction={[Function]}
    addShopToContract={[Function]}
    addShopToStore={[Function]}
    centerPosition={
        Object {
            "lat": 30,
            "lng": 3,
          }
    }
    dispatchDisplayShopWillAppearNoticeAction={[Function]}
    endTransaction={[Function]}
    fetchAll={[Function]}
    goBack={[Function]}
    isTransactionPending={false}
    openNotificationModal={[Function]}
    pendingShop={
        Object {
            "addressString": "40 rue Citadelle Paris",
            "cat": "cool",
            "countryId": "fr",
            "description": "Super Magasin",
            "lat": "30",
            "lng": "3",
            "name": "Magasin",
            "opening": "0000000",
            "postalCode": "75010",
          }
    }
    transactionHash="0x23874grfb2yb38f4h23784"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ShopRecap
          address="40 rue Citadelle Paris"
          addressString="40 rue Citadelle Paris"
          cat="cool"
          countryId="fr"
          description="Super Magasin"
          lat="30"
          licencePrice={null}
          lng="3"
          name="Magasin"
          opening="0000000"
          postalCode="75010"
/>,
        <styled.div>
          <Button
                    disabled={false}
                    fullWidth={false}
                    isSmall={false}
                    onClick={[Function]}
                    theme="primary"
                    width="45%"
          >
                    Submit
          </Button>
          <Button
                    disabled={false}
                    fullWidth={false}
                    isSmall={false}
                    onClick={[Function]}
                    theme="light"
                    width="45%"
          >
                    Edit
          </Button>
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "address": "40 rue Citadelle Paris",
          "addressString": "40 rue Citadelle Paris",
          "cat": "cool",
          "countryId": "fr",
          "description": "Super Magasin",
          "lat": "30",
          "licencePrice": null,
          "lng": "3",
          "name": "Magasin",
          "opening": "0000000",
          "postalCode": "75010",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Button
              disabled={false}
              fullWidth={false}
              isSmall={false}
              onClick={[Function]}
              theme="primary"
              width="45%"
>
              Submit
</Button>,
            <Button
              disabled={false}
              fullWidth={false}
              isSmall={false}
              onClick={[Function]}
              theme="light"
              width="45%"
>
              Edit
</Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Submit",
              "disabled": false,
              "fullWidth": false,
              "isSmall": false,
              "onClick": [Function],
              "theme": "primary",
              "width": "45%",
            },
            "ref": null,
            "rendered": "Submit",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Edit",
              "disabled": false,
              "fullWidth": false,
              "isSmall": false,
              "onClick": [Function],
              "theme": "light",
              "width": "45%",
            },
            "ref": null,
            "rendered": "Edit",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ShopRecap
            address="40 rue Citadelle Paris"
            addressString="40 rue Citadelle Paris"
            cat="cool"
            countryId="fr"
            description="Super Magasin"
            lat="30"
            licencePrice={null}
            lng="3"
            name="Magasin"
            opening="0000000"
            postalCode="75010"
/>,
          <styled.div>
            <Button
                        disabled={false}
                        fullWidth={false}
                        isSmall={false}
                        onClick={[Function]}
                        theme="primary"
                        width="45%"
            >
                        Submit
            </Button>
            <Button
                        disabled={false}
                        fullWidth={false}
                        isSmall={false}
                        onClick={[Function]}
                        theme="light"
                        width="45%"
            >
                        Edit
            </Button>
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "address": "40 rue Citadelle Paris",
            "addressString": "40 rue Citadelle Paris",
            "cat": "cool",
            "countryId": "fr",
            "description": "Super Magasin",
            "lat": "30",
            "licencePrice": null,
            "lng": "3",
            "name": "Magasin",
            "opening": "0000000",
            "postalCode": "75010",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Button
                disabled={false}
                fullWidth={false}
                isSmall={false}
                onClick={[Function]}
                theme="primary"
                width="45%"
>
                Submit
</Button>,
              <Button
                disabled={false}
                fullWidth={false}
                isSmall={false}
                onClick={[Function]}
                theme="light"
                width="45%"
>
                Edit
</Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Submit",
                "disabled": false,
                "fullWidth": false,
                "isSmall": false,
                "onClick": [Function],
                "theme": "primary",
                "width": "45%",
              },
              "ref": null,
              "rendered": "Submit",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Edit",
                "disabled": false,
                "fullWidth": false,
                "isSmall": false,
                "onClick": [Function],
                "theme": "light",
                "width": "45%",
              },
              "ref": null,
              "rendered": "Edit",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
